### PR TITLE
Lyalpha

### PR DESCRIPTION
**Pull request by qyx268**
This should be included in the master now since we've submitted the paper

### CODE REVIEW

File: src/py21cmmc/likelihood.py [hunk 1]
```diff
@@ -185,7 +185,11 @@ def _read_data(self):
                     )
                 )
             else:
-                data.append(dict(np.load(fl, allow_pickle=True)))
+                try:
+                    data.append(dict(np.load(fl, allow_pickle=True)))
+                except TypeError:
+                    # TODO: need to better handle this
+                    data.append(dict(np.load(fl, allow_pickle=True).item()))
```
Comment (steven-murray) [inline src/py21cmmc/likelihood.py]: what's happening here?

**Review by steven-murray — APPROVED** [review]
This looks good. Just one comment to address, and I also noticed there were a bunch of "empty" files... is that expected?

File: src/py21cmmc/likelihood.py (hunk 1 above)
Comment (qyx268) [reply src/py21cmmc/likelihood.py]: hmm... I think it was a numpy version thing that my forest data has to have .item() when load in

**Review by qyx268 — COMMENTED** [review]
the npy files are not empty. They are just not readable by github I think

**Comment by qyx268** [comment]
hmm.. why is travis not checking?

**Comment by qyx268** [comment]
```Builds have been temporarily disabled for public repositories due to a negative credit balance. Please go to the Plan page to replenish your credit balance or alter your Consume paid credits for OSS setting.```

Is this why you moved 21cmfast away from travis? @steven-murray

**Comment by steven-murray** [comment]
Yes indeed it is. Also, Actions are just a bit nicer and more flexible.

Let me know if you want me to do this. I can do it tomorrow.

**Comment by qyx268** [comment]
I've copied/pasted from 21cmfast and updated it accordingly. Could you please take a look and update them if needed? Also, How do we stop requiring Travis test? It seems that we need to send an email to stop the service?

File: src/py21cmmc/core.py [hunk 2]
```diff
@@ -639,3 +651,214 @@ def convert_model_to_mock(self, ctx):
...
+class CoreForest(CoreLightConeModule):
+    r"""A Core Module that produces model effective optical depth at a range of redshifts.
+
+    name : str
+        The name used to match the likelihood
+
+    observation : str
+        The observation that is used to construct the tau_eff statisctic.
+
+    N_realization : int
```
Comment (steven-murray) [inline src/py21cmmc/core.py]: Can we make this `n_realization` (all lower case)?

File: src/py21cmmc/core.py [hunk 3]
```diff
@@ -639,3 +651,214 @@ def convert_model_to_mock(self, ctx):
...
+            or self.observation == "bosman_pessimistic"
+        ):
+            data = np.load(
+                path.join(path.dirname(__file__), "data/Forests/Bosman18/data.npy"),
+                allow_pickle=True,
+            ).item()
+            targets = (data["zs"] > self.redshift[0] - 0.1) * (
+                data["zs"] <= self.redshift[0] + 0.1
+            )
+            self.Nlos = sum(targets)
```
Comment (steven-murray) [inline src/py21cmmc/core.py]: similarly here

File: src/py21cmmc/core.py [hunk 4]
```diff
@@ -639,3 +651,214 @@ def convert_model_to_mock(self, ctx):
...
+            raise ValueError(
+                "You asked for %d realizations, larger than what the box has (Total los / needed los = %d / %d)! Increase HII_DIM!"
+                % (self.N_realization, self.user_params.HII_DIM ** 2, self.Nlos)
+            )
+
+    def setup(self):
+        """Run post-init setup."""
+        CoreBase.setup(self)
+
+    def tau_GP(self, Gamma_bg, Delta, Temp, redshifts):
```
Comment (steven-murray) [inline src/py21cmmc/core.py]: again, here would like all variables to be lower case

File: src/py21cmmc/likelihood.py (hunk 1 above)
Comment (steven-murray) [reply src/py21cmmc/likelihood.py]: are you sure it shouldn't be `.items()`? Is there a test of this?

**Review by steven-murray — COMMENTED** [review]
Just a few stylistic comments, plus a request for a test specifically covering this likelihood (if possible!)

**Comment by steven-murray** [comment]
I've setup the coveralls so that it actually reports a status, and also deactivated the travis checks.

**Comment by qyx268** [comment]
I've updated the variable names. 

For the `.item` issue, maybe I didn't configure `data/Forests/Bosman18/data.npy` correctly? This is what happened if I did `.items()` or don't do `.item()`

image

**Comment by qyx268** [comment]
@steven-murray  can i ask for some help? I don't think I'm understaning the test correctly. 

So I had to recreate a lc object and context because the forest data needs to reach z<6. I think lc_core_lowz_ctx should have the forest core in the context as I've added it into the cores when building the chain. However, in the test function, it seems that the ctx actually didn't run any simulation(?). Then I get an error reporting tau_eff is None.

How do I make sure the forest core is called and the modelled forest data is generated?

**Comment by steven-murray** [comment]
Hi @qyx268. I'll give it a look -- I can't answer the question off the top of my head. Hopefully can get to it today, but I am pretty packed :/

**Comment by steven-murray** [comment]
Hi @qyx268. Firstly concerning the numpy load. Can you try this:

image

I can't reproduce your errors, but I'll bet it's how you saved the data.

**Comment by steven-murray** [comment]
I updated the test. I think the problem was that you were using the other `Core` to build the model data. You can use the `LikelihoodComputationChain` itself to build the model data, which will use all the available cores.

**Comment by qyx268** [comment]
@steven-murray  thx for the help. For the coverall test, is it just because we didn't have one on master to compare against?

**Comment by qyx268** [comment]
it's actually 81% coverage for the master...  google saied i need to close and reopen

**Pull request merged**